include/envoy/stats/timespan.h

```diff
@@ -24,35 +24,61 @@ class CompletableTimespan {
 };
 
 /**
- * An individual timespan that flushes its measured value in time unit (e.g
- * std::chrono::milliseconds). The initial time is captured on construction. A timespan must be
+ * An individual timespan that flushes its measured value to the histogram on completion.
+ * The start time is captured on construction. The timespan must be
  * completed via complete() for it to be stored. If the timespan is deleted this will be treated as
- * a cancellation.
+ * a cancellation. The target histogram must represent a quantity of time.
  */
-template <class TimeUnit> class TimespanWithUnit : public CompletableTimespan {
+class HistogramCompletableTimespan : public CompletableTimespan {
 public:
-  TimespanWithUnit(Histogram& histogram, TimeSource& time_source)
-      : time_source_(time_source), histogram_(histogram), start_(time_source.monotonicTime()) {}
+  HistogramCompletableTimespan(Histogram& histogram, TimeSource& time_source)
+      : time_source_(time_source), histogram_(histogram), start_(time_source.monotonicTime()) {
+    switch (histogram.unit()) {
+    case Histogram::Unit::Unspecified:
+    case Histogram::Unit::Bytes:
+      ASSERT(0);
```

Comment:
I like the change to templatize only the method. Can we then move rest of the impls into an _impl.cc file?

Reply:
Fixed.